proposals/0000-split-wunused-imports.md

```diff
@@ -0,0 +1,230 @@
+
+## Motivation
+
+`-Wunused-imports` has long been a thorn in the side of people who want both:
+* to not have unused imports (in the sense described in the [proposed change specification section](#proposed-change-specification))
+* to compile warning-free across different versions of their dependencies
+
+The most often case where this comes up is with `Prelude` and other widely used modules in `base`.
+In particular, when `Prelude` gets a new reexport, which previously only lived in another module from `base`,
+users of `-Wall` suddenly get new warnings, causing not only noise, but crucially breakage for users of `-Wall -Werror`.
+
+This effectively reduces the usability of the `-Wunused-imports` warning,
+because it forces library authors (who in general are doing this work for free) who want to enforce code quality (removing actual unused imports and potentially even unused dependencies)
+to also have to do extra work, with questionable benefits.
+
+This not only a theoretic concern:
+
+* Additional work was required in the implementation of the [export `liftA2` proposal](https://github.com/haskell/core-libraries-committee/issues/50):
+    * effort in `ghc` itself to maintain warning-freeness
+    * additional non-trivial sections in [the migration guide](https://github.com/haskell/core-libraries-committee/blob/main/guides/export-lifta2-prelude.md)
+    * two additional non-trivial(requiring `CPP` or turning off the warning entirely) merge requests, to `Cabal-syntax` and `containers` respectively
+
+* The implementation of the AMP proposal required very similar additional work to the `liftA2` case above, see
+    * the concerns raised in https://www.yesodweb.com/blog/2016/05/are-unused-import-warnings-harmful
+    * the migration guide for ghc 7.10 - https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/7.10#ghc-says-the-import-of-is-redundant
+
+* This is listed as a "bad part" of Haskell - https://www.snoyman.com/blog/2020/11/haskell-bad-parts-2/
+    * in particular, the article mentions the re-export of `(<>)` from `Prelude` as requiring the exact same `CPP` workarounds as already mentioned
+
+* In the [sized integer types proposal](https://github.com/haskell/core-libraries-committee/issues/156), a critical counterpoint is introducing new emissions of `-Wunused-imports`
+
+* The [export `foldl'` proposal](https://github.com/haskell/core-libraries-committee/issues/167) seems like it would encounter the same implementation issues as the export `liftA2` proposal, and similar concerns to the sized integer types proposal
+
+## Proposed Change Specification
+
+An "unused import" looks like this:
+```haskell
+import X (f)
+import Z
+...neither f, nor anything from Z is used, hence they are both unused imports...
+```
+A "duplicate import" looks like this:
+```haskell
+import X
+import X (f)
+...f is used here...
+```
+Here, one of the imports is currently marked as unused.
```

Comment:
The proposed spec change only talks about imports of the same module, but (based on the example).it should also apply when importing different modules that re-export things, right? I think a more precise specification wouldn't hurt.

Also, one can argue that one really only want to change this for _different_ modules, and still always warn about ”blatant duplication” of the same module import, which is neither good style, not does it actually help with the problem we want to solve.

Thinking this to it's conclusion might result in a design much like Ben's design at https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/compiler/relaxed-unused-imports.

Reply:
Yep, you're right, I should extend the proposed specification some, thanks!

Reply:
I'm resolving this to keep noisiness to a minimum, please add another comment if you feel that the proposed change specification still needs work.